### ISSUE TITLE
For generic structs, apply type bounds in impl rather than requiring on base

### ISSUE THREAD

**Issue by TedDriggs** [colin-kiegel/rust-derive-builder#91]
If you run `cargo expand` on this code:

```rust
#[derive(Clone)]
pub struct Foo<T>(T);

fn main() {}
```

You'll see that the generated `Clone` impl has a type bound requiring that `T: ::std::clone::Clone`:

```rust
#![feature(prelude_import)]
#![no_std]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std as std;
pub struct Foo<T>(T);
#[automatically_derived]
#[allow(unused_qualifications)]
impl<T: ::std::clone::Clone> ::std::clone::Clone for Foo<T> {
    #[inline]
    fn clone(&self) -> Foo<T> {
        match *self {
            Foo(ref __self_0_0) => Foo(::std::clone::Clone::clone(&(*__self_0_0))),
        }
    }
}

fn main() {}
```

This may be possible for the generated builder impl as well: Inherent impl blocks are allowed to have trait bounds beyond what is declared on the base type. Doing so would better align to the standard library and would be slightly more ergonomic.

# Implementation Considerations
1. The `Clone` type bound should not be emitted when using the `owned` builder pattern.
1. The `Default` type bound should not be emitted for generic fields with an explicit default expression.

_Update:_ Turns out that the compiler is perfectly happy with duplicate type bounds:

```rust
// stuff elided...

pub struct Foo<T: ::std::clone::Clone>(T);
#[automatically_derived]
#[allow(unused_qualifications)]
impl<T: ::std::clone::Clone + ::std::clone::Clone> ::std::clone::Clone for Foo<T> {
    #[inline]
    fn clone(&self) -> Foo<T> {
        match *self {
            Foo(ref __self_0_0) => Foo(::std::clone::Clone::clone(&(*__self_0_0))),
        }
    }
}
```

That makes this quite a bit easier.

_Update 2:_ The compiler makes another simplifying assertion:
It requires all generic type params implement `Clone`, rather than all field members' types.

```rust
#[derive(Clone)]
pub struct Lorem<T, U> {
    ipsum: T,
    sit: bool,
    dolor: Doubler<U>,
}

pub struct Doubler<T>(T);

impl Clone for Doubler<SomeUnstableThing> {
    fn clone(&self) -> Self {
        Doubler(SomeUnstableThing::new())
    }
}
```

In this code, it requires `U: ::std::clone::Clone`, rather than adding a `where` clause which would constrain the `Doubler`. This is a little surprising to me, but I suspect it makes life easier since it means there isn't a need to determine which fields reference the various generic types.

**Comment by TedDriggs** [comment]
See rust-lang/rust#26925 for rationale behind behavior of built-in traits. I'm comfortable with that behavior for `Clone` but not for `Default` so my fix will be limited to the Clone trait.

**Comment by colin-kiegel** [comment]
@TedDriggs Ok, cool. :-)

We don't ever emit `T: Default` bounds for generic builders. This means that we are not be to compile something like `foo: T = Default::default()` for a generic parameter `T`. I opened a ticket #93. ^^

**Comment by colin-kiegel** [comment]
@TedDriggs Can we close this and track the remaining `T: Default` problem in #93?

**Comment by TedDriggs** [comment]
My inclination is to close it and discuss `T: Default` in issue #93. There's a "milestone" field for issues that might be useful for tracking that this hasn't shipped yet.

**Comment by colin-kiegel** [comment]
Released with 0.4.7.
